Application Commands Documentation (#309)
nextcord/nextcord: docs/interactions.rst

:orphan:

.. currentmodule:: nextcord

.. _interactions:
   
Slash Commands
==============

> Since Discord has added interaction commands a feature with many possibilities, we at Nextcord have decided to add them to our fleet of features.

Reviewer: Not a fan of this line, it feels *too* 'friendly'/'joky'. It's expected of us as a lib to implement Discord's API feature's, but this sounds as if it was just a general feeling that we should

Author: Yeah I can’t find any good replacement to this line

Reviewer 2: Would it be better to just not include any description here?

> This doc will explain the innerworkings and how to use interaction commmands.

Reviewer: ```suggestion
This document will explain the inner workings and how to use interaction commands.
```

> We suggest you learn how to make regular commands before looking through here, we suggest looking at the :doc:`quickstart`

Reviewer: ```suggestion
We suggest you learn how to make regular commands before looking through here, we suggest looking at the main :documentation:`quickstart` for commands.
```

> We suggest you learn how to make regular commands before looking through here, we suggest looking at the :doc:`quickstart`

Reviewer 2: no that's stupid?

How To Make A Simple Slash Command
----------------------------------

> This right here is a simple ping pong command made with Nextcord's slash feature.

Reviewer: ```suggestion
The following example is a simple ping/pong command using Nextcord's slash commands:
```

.. code-block:: python3

    @client.slash_command()
    async def ping(interaction):
        await interaction.response.send_message("Pong!")
        

> The way it works is that you use the :meth:`~Client.slash_command` function to interact with discord's application commands endpoint.

Reviewer: ```suggestion
The way it works is by using:meth:`~Client.slash_command` decorator to add a slash command for the bot to register with Discord.
```

Reviewer 2: @vcokltfre shouldnt it be *the* @slash_command deco...

> ``guild_ids`` is used to limit the guilds that the slash command is available to. This is also useful for testing, as global slash commands can take up to an hour to register.

Reviewer 2: ```suggestion
``guild_ids`` is used to limit the guilds that the slash command is available to. This is useful for testing as global slash commands can take up to an hour to register.
```

Example:

.. code-block:: python3
      
    @client.slash_command(guild_ids=[id1, id2])

How To Use Sub-Commands
-----------------------

> The way sub-commands work is that you make a normal slash command that will never be called, and then make the sub-commands and have them do the work of real slash commands. There is no difference with slash commands and sub-commands. The only thing you will need to change is functions.

Reviewer 2: I generally dislike how this is worded

As shown in the demonstration below you make a main slash command or a dummy slash command and build sub-commands off it.

.. code-block:: python3

        @client.slash_command(guild_ids=[...])  # Making the command and limiting the guilds

>         async def my_main_command(interaction):  

Reviewer: ```suggestion
        async def my_main_command(interaction: Interaction):  
```
Typehinting should be added here for consistency.

            ...

        @my_main_command.subcommand()  # Identifying The Sub-Command

>         async def subcommand_one(interaction):  # Making The Sub Command Name And Passing Through Interaction

Reviewer: ```suggestion
        async def subcommand_one(
            interaction: Interaction
        ):  # Making The Sub Command Name And Passing Through Interaction
```
Typehinting should be added here for consistency.

        await interaction.response.send_message(
            "This is subcommand 1!"

>             )  # Sending A Response

Reviewer: ```suggestion
        )  # Sending A Response
```

        # Identifying The Sub-Command And Adding A Descripton
        @my_main_command.subcommand(description="Aha! Another subcommand")  
        async def subcommand_two(interaction: Interaction):  # Passing in interaction

        await interaction.response.send_message(
            "This is subcommand 2!"

>             )       # Responding With The Args/Fields

Reviewer: ```suggestion
            )  # Responding With The Args/Fields
```

>             )       # Responding With The Args/Fields

Reviewer: ```suggestion
        )       # Responding With The Args/Fields
```

> Fields And Requirements

Reviewer: Nothing in the following section is named a 'field', I'm an advanced user and I have a difficult time trying to understand what is meant by half of this. I'd suggest rewriting this portion to reflect the naming in the library itself.

-----------------------
Fields are meant to facilitate an easier way to fill info, letting people using your slash command get a different response for each answer.

Nextcord's implementation of slash commands has fields and is very simple. in the example below is a field.

.. code-block:: python3
     
     @client.slash_command()
     async def help(
         interaction: Interaction,
         setting: str = SlashOption(name="settings", description="Configure Your Settings")
     ):
         if setting == "music":
             await interaction.response.send_message("Sorry we don't have PyNaCl installed currently")
         elif setting == "moderation":
             await interaction.response.send_message("Moderation?")
         else:

>              await interaction.response.send_message("Odd, I don't know that setting")

Reviewer: Surely this would be a better place to add options into the example, so that music and moderation are the only options given to the user? This would ensure people are using the right tools from the job right away, rather than learning about them much later when they already have code with unnecessary handlers like this in production.

How To Make Slash Commands In Cogs
----------------------------------

> Show below is an example of a simple command running in a cog.

Reviewer: ```suggestion
Shown below is an example of a simple command running in a cog:
```

.. code-block:: python3

    class ExampleCog(commands.Cog):
        def __init__(self):
            self.count = 0

        @nextcord.slash_command()

>         async def slash_example_cog_command(self, interaction):

Reviewer: ```suggestion
        async def slash_example_cog_command(self, interaction: Interaction):
```
Typehinting should be added here for consistency.

>             await interaction.response.send_message("Hello i am a slash command in a cog!")

Reviewer: ```suggestion
            await interaction.response.send_message("Hello I am a slash command in a cog!")
```

> The example shown above responds to a user when they do a slash command. It is identical to a normal slash command and to normal commands in general.

Reviewer: ```suggestion
The example shown above responds to a user when they do a slash command. Its function is the same as a slash command on the client, adjusted to work in a class, only its decorator is different.
```

How To Make Context Menu Commands
---------------------------------

> Context Menu commands display commands on a menu of a message/user.

Reviewer: ```suggestion
Context Menu commands display commands on the right-click menu of a message/user.
```

User Commands
~~~~~~~~~~~~~~

> What we show below just dumps the user's username but should be enough to get the point.

Reviewer: Too informal for documentation of this type.
```suggestion
The following code simply prints out the name of the member it was performed on, but can be used for more complex actions too:
```

.. code-block:: python3

    @client.user_command()

>     async def hello(interaction, member):

Reviewer: ```suggestion
    async def hello(interaction: Interaction, member: Member):
```
Typehinting should be added here for consistency.

        await interaction.response.send_message(f"Hello! {member}")

Message Commands
~~~~~~~~~~~~~~~~~

> Below is a simple example of message command that says the message given.

Reviewer: ```suggestion
The following is a simple example of a message command which sends the content of the target message as an ephemeral response to the user who invoked it.
```

.. code-block:: python3

    @client.message_command()

>     async def say(interaction, message: Message):

Reviewer: ```suggestion
    async def say(interaction: Interaction, message: Message):
```
Typehinting should be added here for consistency.

>         await interaction.response.send_message(f"{message}")

Reviewer: ```suggestion
        await interaction.response.send_message(message.content, ephemeral=True)
```

:orphan:

.. currentmodule:: nextcord

.. _interactions:
   
Slash Commands
==============

Since Discord has added interaction commands a feature with many possibilities, we at Nextcord have decided to add them to our fleet of features.

This doc will explain the innerworkings and how to use interaction commmands.

We suggest you learn how to make regular commands before looking through here, we suggest looking at the :doc:`quickstart`

How To Make A Simple Slash Command
----------------------------------

> This right here is a simple ping pong command made with Nextcords slash feature.

Reviewer 2: ```suggestion
This right here is a simple ping pong command made with Nextcord's slash feature.
```

.. code-block:: python3

    @bot.slash_command(name="ping")
    async def ping(interaction):
        await interaction.response.send_message("Pong!")
        

> The way it works is that you use the :meth:`~Client.slash_command` function to interact with the Discord API. The ``name`` parameter is the name of your slash command.

Reviewer 2: > function to interact with the Discord API.
This is awfully worded imo, also unnecessary as the entire lib is about interacting with the discord api?

``guild_ids`` is used to limit the guilds that the slash command is available to. This is also useful for testing, as global slash commands can take up to an hour to register.

Example:

.. code-block:: python3
      

>     @bot.slash_command(guild_ids=[id1, id2])

Reviewer 2: why are we using bot instead of client here?

How To Use Sub-Commands
-----------------------

The way sub-commands work is that you make a normal slash command that will never be called, and then make the sub-commands and have them do the work of real slash commands. There is no difference with slash commands and sub-commands. The only thing you will need to change is functions.

As shown in the demonstration below you make a main slash command or a dummy slash command and build sub-commands off it.

.. code-block:: python3

        @bot.slash_command(guild_ids=[...])  # Making the command and limiting the guilds

>         async def main(interaction):  

Reviewer 2: This could possibly be renamed to something slightly more meaningful?

>         # passing through interaction and indentifing the sub-command group name
>         await interaction.response.send_message(
>         "This will never get called if this has subcommands."
>             )  # a function never called

Reviewer 2: ```suggestion
        ...
```

        @main.subcommand()  # Identifying The Sub-Command
        async def sub1(interaction):  # Making The Sub Command Name And Passing Through Interaction
        await interaction.response.send_message(
            "This is subcommand 1!"
            )  # Sending A Response

        @main.subcommand(description="Aha! Another subcommand")  

>         # Identifying The Sub-Command And Adding A Descripton

Reviewer 2: The comment should be above the deco

        async def subcommand_two(interaction: Interaction):  # Passing in interaction

        await interaction.response.send_message(
            "This is subcommand 2!"
            )       # Responding With The Args/Fields
        
Fields And Requirements
-----------------------
Fields are meant to facilitate an easier way to fill info, letting people using your slash command get a different response for each answer.

Nextcord's implementation of slash commands has fields and is very simple. in the example below is a field.

.. code-block:: python3
     
     @bot.slash_command(name="help")
     async def help(
         interaction: Interaction,
         setting: str = SlashOption(name="settings", description="Configure Your Settings")
     ):
         if setting == "music":

>              await interaction.response.send_message(f"Sorry we aren't connected to lavalink currently")

Reviewer 2: references to other libraries is probably not a good idea. Also unnecessary f string

         elif setting == "moderation":
             await interaction.response.send_message(f"Moderation?")
         else:
             await interaction.response.send_message("Odd, I don't know that setting")

How To Make Slash Commands In Cogs
----------------------------------
Show below is an example of a simple command running in a cog.

.. code-block:: python3

    class ExampleCog(commands.Cog):
        def __init__(self):
            self.count = 0

>         @slash_command(name="cogexample")

Reviewer 2: this should also include the nextcord namespace (also imports are missing?)

        async def slash_example_cog_command(self, interaction):
            await interaction.response.send_message("Hello i am a slash command in a cog!")

The example shown above responds to a user when they do a slash command. It is identical to a normal slash command and to normal commands in general.

How To Make Context Menu Commands
---------------------------------
Context Menu commands display commands on a menu of a message/user.

User Commands
~~~~~~~~~~~~~~
What you see below is a example of a simple user command.

> It is a user dump command that dumps user data.

Reviewer 2: This could probably be simplified a bit and like just give the users username? Also possibly say something about thats the command when you use the apps menu on a user?

.. code-block:: python3

    @bot.user_command(name="dump")
    async def userdump(interaction, member):
        await interaction.response.send_message(f"Member: {member}, Data Dump: {interaction.data}")

Message Commands
~~~~~~~~~~~~~~~~~

> What you see below is a example of a simple message command, 

Reviewer 2: Same here

.. code-block:: python3

    @bot.slash_command(name="ping")
    async def ping(interaction):
        await interaction.response.send_message("Pong!")
        
The way it works is that you use the :meth:`~Client.slash_command` function to interact with the Discord API. The ``name`` parameter is the name of your slash command.

``guild_ids`` is used to limit the guilds that the slash command is available to. This is also useful for testing, as global slash commands can take up to an hour to register.

Example:

.. code-block:: python3
      
    @bot.slash_command(guild_ids=[id1, id2])

How To Use Sub-Commands
-----------------------

The way sub-commands work is that you make a normal slash command that will never be called, and then make the sub-commands and have them do the work of real slash commands. There is no difference with slash commands and sub-commands. The only thing you will need to change is functions.

As shown in the demonstration below you make a main slash command or a dummy slash command and build sub-commands off it.

.. code-block:: python3

        @bot.slash_command(guild_ids=[...])  # Making the command and limiting the guilds

>         async def main(

Reviewer 3: commenting here for vincent

                interaction,
            ):  # passing through interaction and indentifing the sub-command group name
        await interaction.response.send_message(
        "This will never get called if this has subcommands."
            )  # a function never called

        @main.subcommand()  # Identifying The Sub-Command
        async def sub1(interaction):  # Making The Sub Command Name And Passing Through Interaction
        await interaction.response.send_message(
            "This is subcommand 1!"
            )  # Sending A Response

>         @main.subcommand(

Reviewer 3: here too

Example:

.. code-block:: python3
      
    @bot.slash_command(guild_ids=[id1, id2])

How To Use Sub-Commands
-----------------------

The way sub-commands work is that you make a normal slash command that will never be called, and then make the sub-commands and have them do the work of real slash commands. There is no difference with slash commands and sub-commands. The only thing you will need to change is functions.

As shown in the demonstration below you make a main slash command or a dummy slash command and build sub-commands off it.

.. code-block:: python3

        @bot.slash_command(guild_ids=[...])  # Making the command and limiting the guilds
        async def main(
                interaction,
            ):  # passing through interaction and indentifing the sub-command group name
        await interaction.response.send_message(
        "This will never get called if this has subcommands."
            )  # a function never called

        @main.subcommand()  # Identifying The Sub-Command
        async def sub1(
                interaction,
            ):  # Making The Sub Command Name And Passing Through Interaction

>         await interaction.response.send_message(

Reviewer 3: why are these brackets like
```py
this(
a
)
```
when its so small it can be done like
```py
this(a)
```
